reference/behavior_test/establishing_decorator_pattern.rst

```diff
@@ -0,0 +1,369 @@
+
+Gherkin feature :
+
+.. code-block:: gherkin
+
+    Scenario: View only the mass edit operations I have permissions on
+        Given I edit the "Catalog manager" Role # Change the current page and ask to Selenium to open it in the navigator.
+        And I visit the "Permissions" tab
+        And I grant rights to group Products # Here we do the action to revoke rights.
+        And I revoke rights to resource Change state of product
+        And I save the Role
+        Then I should not see the text "There are unsaved changes."
+        When I am on the products page
+        And I mass-edit products kickers and hiking_shoes
+        And I should not see "Change status (enable / disable)"
+
+Here we describe what happens if we revoke rights to change state of product to an user in the catalog manager role group.
+
+Context
+_______
+
+Navigate And Changing The Current Page
+**************************************
+
+*Given I edit the "Catalog manager" Role* With this line we change the current page to be on "Role edit".
+
+.. code-block:: php
+
+    // \Pim\Behat\Context\NavigationContext
+    /**
+     * @param string $identifier
+     * @param string $page
+     *
+     * Gherkin match this regexp
+     * @Given /^I edit the "([^"]*)" (\w+)$/
+     */
+    public function iAmOnTheEntityEditPage($identifier, $page) // $identifier = 'Catalog manager', $page = 'Role'
+    {
+        $page   = ucfirst($page);
+        $getter = sprintf('get%s', $page);
+        $entity = $this->getFixturesContext()->$getter($identifier);
+        $this->openPage(sprintf('%s edit', $page), ['id' => $entity->getId()]);
+    }
+
+    /**
+     * @param string $pageName
+     * @param array  $options
+     *
+     * @return \SensioLabs\Behat\PageObjectExtension\PageObject\Page
+     */
+    public function openPage($pageName, array $options = [])
+    {
+        // "Role edit" is now our current page. The current Page is the one we are working on.
+        $this->currentPage = $pageName;
+
+        // Ask to Selenium to open "/user/role/update/{id}" in the
+        // navigator thank to $path containing in "Context\Page\Role\Edit".
+        $page = $this->getCurrentPage()->open($options);
+
+        $this->wait();
+
+        return $page;
+    }
+
+Selenium will know which url to go thank to `the protected $path attribute
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Context/Page/Role/Edit.php#L19>`_
+containing in *Context\\Page\\Role\\Edit*.
+
+You may remark that this Gherkin method still use the old implementation of Behat in Akeneo PIM.
+
+Call Pages Actions
+******************
+
+*And I grant rights to group Products* Now, let's jump in Decorators !
+
+All our new contexts are in the namespace *\\Pim\\Behat\\Context\\Domain* which is classified by domain. Each domain is
+linked to a page in Akeneo PIM. That's why you will find in this namespace the same architecture than our main menu.
```

Comment:
That's why, in this namespace, you will find the same architecture as in our main menu.